app/src/main/java/com/OxGames/Pluvia/service/SteamUnifiedFriends.kt

```
@@ -0,0 +1,396 @@
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+// References:
+// https://github.com/marwaniaaj/RichLinksJetpackCompose/tree/main
+// https://blog.stackademic.com/rick-link-representation-in-jetpack-compose-d33956e8719e
+// https://github.com/lukasroberts/AndroidLinkView
+// https://github.com/android/compose-samples/tree/main/Jetchat
+// https://github.com/LossyDragon/Vapulla
+
+// TODO
+//  Implement Reactions
+//  OfflineMessageNotificationCallback ?
+//  FriendMsgEchoCallback ?
+//  EmoticonListCallback ?
+
+typealias AckMessageNotification = SteammessagesFriendmessagesSteamclient.CFriendMessages_AckMessage_Notification.Builder
+typealias IncomingMessageNotification = SteammessagesFriendmessagesSteamclient.CFriendMessages_IncomingMessage_Notification.Builder
+typealias FriendNicknameChanged = SteammessagesPlayerSteamclient.CPlayer_FriendNicknameChanged_Notification.Builder
+
+class SteamUnifiedFriends(
+    private val service: SteamService,
+) : AutoCloseable {
+
+    private var unifiedMessages: SteamUnifiedMessages? = null
+
+    private var chat: Chat? = null
+
+    private var player: Player? = null
+
+    private var friendMessages: FriendMessages? = null
+
+    init {
+        unifiedMessages = service.steamClient!!.getHandler<SteamUnifiedMessages>()
+
+        chat = unifiedMessages!!.createService(Chat::class.java)
+
+        player = unifiedMessages!!.createService(Player::class.java)
+
+        friendMessages = unifiedMessages!!.createService(FriendMessages::class.java)
+
+        with(service.callbackManager!!) {
+            with(service.callbackSubscriptions) {
+                add(subscribeServiceNotification<FriendMessagesClient, IncomingMessageNotification>(::onIncomingMessage))
+                add(subscribeServiceNotification<FriendMessages, AckMessageNotification>(::onAckMessage))
+                add(subscribeServiceNotification<PlayerClient, FriendNicknameChanged>(::onNickNameChanged))
+            }
+        }
+    }
```

**suggestion: (untested)** use null-safe calls instead of !!

```kotlin
service.steamClient?.getHandler<SteamUnifiedMessages>()?.let {
    unifiedMessages = it
    chat = it.createService(Chat::class.java)
    player = it.createService(Player::class.java)
    friendMessages = it.createService(FriendMessages::class.java)
}

service.callbackManager?.let { callbackManager ->
    with(callbackManager) {
        with(service.callbackSubscriptions) {
            add(subscribeServiceNotification<FriendMessagesClient, IncomingMessageNotification>(::onIncomingMessage))
            add(subscribeServiceNotification<FriendMessages, AckMessageNotification>(::onAckMessage))
            add(subscribeServiceNotification<PlayerClient, FriendNicknameChanged>(::onNickNameChanged))
        }
    }
}
```

---

That usually is the safe approach, but I'd rather let it crash if something was to fail. The current code has no way to handle these initializers if they fail to load.